tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs

```diff
@@ -0,0 +1,344 @@
+                    {
+                        if (charArray is null)
+                        {
+                            charArray = value.ToCharArray();
+                        }
+
+                        charArray[i] = lowArray[0];
+                        c = lowArray[0];
+                    }
+                }
+
+                if (char.IsLetter(c))
+                {
+                    chars++;
+                }
+                else if (chars == 0)
+                {
+                    trim = i + jump;
+                    goto end;
+                }
+                else if (char.IsDigit(c) || c == '.' || c == '/' || c == '-')
+                {
+                    chars++;
+                }
+                else
+                {
+                    if (cuts is null)
+                    {
+                        cuts = new List<int[]>();
+                    }
+
+                    var n = cuts.Count;
+                    if (n > 0 && cuts[n - 1][1] >= i)
+                    {
+                        cuts[n - 1][1] += jump;
+                    }
+                    else
+                    {
+                        cuts.Add(new int[] { i, i + jump });
+                    }
+                }
+
+            end:
+                if (i + jump >= 2 * MaxTagLength)
+                {
+                    i++;
+                    break;
+                }
+
+                if (chars >= MaxTagLength)
+                {
+                    i++;
+                    break;
+                }
+            }
+
+            i--;
+
+            if (cuts is null || cuts.Count == 0)
+            {
+                if (charArray is null)
+                {
+                    return value.Substring(trim, i + jump - trim);
+                }
+
+                return new string(charArray, trim, i + jump - trim);
+            }
+
+            charArray ??= value.ToCharArray();
+            var segment = new ArraySlice<char>(charArray, trim, i + jump - trim);
+            int delta = trim;
+            foreach (var cut in cuts)
+            {
+                int start = cut[0] - delta;
+                int end = cut[1] - delta;
+
+                if (end >= segment.Count)
+                {
+                    segment = segment.Slice(0, start);
+                    break;
+                }
+
+                segment[start] = '_';
+                if (end - start == 1)
+                {
+                    continue;
+                }
+
+                segment.Slice(end).CopyTo(segment.Slice(start + 1));
+                segment = segment.Slice(0, segment.Count - (end - start) + 1);
+                delta += cut[1] - cut[0] - 1;
+            }
+
+            return new string(segment.Array, segment.Offset, segment.Count);
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/traceutil/normalize.go#L208-L211
```

Comment:
Could you update these comments with SHA permalinks?

Reply:
done in [b28b71a](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/b28b71a6f82bbfa8424589a1b12d3c572d625629)